var/spack/repos/builtin/packages/py-h5py/package.py

```diff
@@ -44,12 +50,26 @@ class PyH5py(PythonPackage):
     depends_on('mpi', when='+mpi')
     depends_on('py-mpi4py', when='+mpi', type=('build', 'run'))
 
+    # Conflicts
+    conflicts('^hdf5@1.12.0:', when='@:2.99')
+
     phases = ['configure', 'install']
 
+    @when('@:2.99')
     def configure(self, spec, prefix):
         self.setup_py('configure', '--hdf5={0}'.format(spec['hdf5'].prefix),
                       '--hdf5-version={0}'.format(spec['hdf5'].version))
 
         if '+mpi' in spec:
             env['CC'] = spec['mpi'].mpicc
             self.setup_py('configure', '--mpi')
+
+    @when('@3.0.0:')
+    def configure(self, spec, prefix):
+        if '+mpi' in spec:
+            env['CC'] = spec['mpi'].mpicc
+            env['HDF5_MPI'] = 'ON'
+            env['HDF5_DIR'] = spec['hdf5'].prefix
```

Comment:
Since this stuff is common to both, let's put it in a separate `setup_build_environment` method.

Reply:
Once that's moved to `setup_build_environment`, how do I specify that the `configure` step should only be run for `@:2.99` ? 
(I can't add a `@when` decorator to the `phases = ...` line since it's not a function and `configure` isn't a base method for the [`PythonPackage`](https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#phases) class. )

Reply:
```python
@property
def phases(self):
    return ['install'] if self.spec.satisfies('@3:') else ['configure', 'install']
```
You could also see if it's possible to run `['build', 'install']` like other `PythonPackage`s.

Reply:
Thanks for the explanation!

Reply:
@adamjstewart Maybe `phases` should just be a method, not a property? The unit tests are failing:
```
==> Error: 'property' object is not iterable
```

Reply:
@alalazo why doesn't this work?

Reply:
If I recall correctly `phases` (as a list of strings) is used by meta-classes to construct and attach more complex descriptor objects - i.e. it's used in a context where a class is being defined, not in a context where an object of that class is already instantiated. In the original implementation (2016?) it was supposed to be a fixed value for build-systems, not something that each package could change dynamically.

Reply:
In that case I guess we'll have to either create a dummy phase for older versions:
```python
@when('@:2')
def configure(self, spec, prefix):
    pass
```
or turn it into a before-install step:
```python
@run_before('install')
def configure(self):
    if self.spec.satisfies('@:2'):
        ...
```

Reply:
The second option looks good to me, maybe with a more specific name:
```python
@run_before('install')
def setup_py_configure(self):
    if self.spec.satisfies('@:2'):
        ...
```
to avoid confusion with the configure phase of other build systems.

Reply:
Meh, I think `configure` makes more sense. Every phase in PythonPackage is named after the arg passed to setup.py.